test/core/pipeline/features/test_run.py

```diff
@@ -2224,3 +2280,498 @@ def that_has_a_string_variadic_component():
+                self.idx += 1
+
+            return {"replies": replies}
+
+
+    code_prompt_template = """
+Generate code to solve the task: {{ task }}
+
+{% if feedback %}
+Here is your initial attempt and some feedback:
+{{ feedback }}
+{% endif %}
+    """
+
+    feedback_prompt_template = """
+Check if this code is valid and can run: {{ code[0] }}
+Return "PASS" if it passes and "FAIL" if it fails.
+Provide additional feedback on why it fails.
+    """
+
+    code_llm = FixedGenerator(replies=["invalid code", "valid code"])
+    code_prompt = PromptBuilder(template=code_prompt_template)
+
+    feedback_llm = FixedGenerator(replies=["FAIL", "PASS"])
+    feedback_prompt = PromptBuilder(template=feedback_prompt_template)
+
+    routes = [
+        {
+            "condition": "{{ 'FAIL' in replies[0] }}",
+            "output": "{{ replies[0] }}",
+            "output_name": "fail",
+            "output_type": str,
+        },
+        {
+            "condition": "{{ 'PASS' in replies[0] }}",
+            "output": "{{ code }}",
+            "output_name": "pass",
+            "output_type": List[str],
+        },
+    ]
+
+    router = ConditionalRouter(routes=routes)
+
+    concatenator = OutputAdapter(template="{{current_prompt[0] + '\n' + feedback[0]}}", output_type=str)
+
+    answer_builder = AnswerBuilder()
+
+    pp = Pipeline(max_runs_per_component=100)
+
+    pp.add_component("code_llm", code_llm)
+    pp.add_component("code_prompt", code_prompt)
+    pp.add_component("feedback_prompt", feedback_prompt)
+    pp.add_component("feedback_llm", feedback_llm)
+    pp.add_component("router", router)
+    pp.add_component("concatenator", concatenator)
+    pp.add_component("answer_builder", answer_builder)
+
+    pp.connect("code_prompt.prompt", "code_llm.prompt")
+    pp.connect("code_llm.replies", "feedback_prompt.code")
+    pp.connect("feedback_llm.replies", "router.replies")
+    pp.connect("router.fail", "concatenator.feedback")
+    pp.connect("router.pass", "answer_builder.replies")
+    pp.connect("code_llm.replies", "router.code")
+    pp.connect("feedback_prompt.prompt", "feedback_llm.prompt")
+    pp.connect("code_llm.replies", "concatenator.current_prompt")
+    pp.connect("concatenator.output", "code_prompt.feedback")
+
+
+
+
+    task = "Generate code to generate christmas ascii-art"
+
+    return (
+        pp,
+        [
+            PipelineRunData(
+                inputs={"code_prompt": {"task": task}, "answer_builder": {"query": task}},
+                expected_outputs={
+                    "answer_builder": {
+                        "answers": [GeneratedAnswer(data="valid code", query=task, documents=[])]
+                    }
+                },
+                expected_run_order=[
+                    'code_prompt', 'code_llm', 'feedback_prompt', 'feedback_llm', 'router', 'concatenator',
+                    'code_prompt', 'code_llm', 'feedback_prompt', 'feedback_llm', 'router', 'answer_builder'],
```

Comment:
This test is very similar to the one above but without user input to the second `PromptBuilder`. The test will fail because `feedback_prompt` and `feedback_llm` are executed a third time. `feedback_prompt` does not receive any inputs but still executes.